Subject: Correct cpu affinity for dlpar added cpus
From: nfont@linux.vnet.ibm.com 
Patch-mainline: v4.1-rc2
References: bsc#928970
Git-commit: f32393c943e297b8ae180c8f83d81a156c7d0412

The incorrect ordering of operations during cpu dlpar causes the affinity
of cpus being added to be invalid. Phyp does not assign affinity information
for a cpu until the rtas set-indicator calls are made to set the isolation
and allocation state. In the current code we call rtas configure-connector
before making the set-indicator calls which results in invalid data in the
ibm,associativity property for the cpu we're adding.

This patch corrects the order of operations to make the set-indicator
calls (done in acquire_drc) before calling configure-connector.

Signed-off-by: Nathan Fontenot <nfont at linux.vnet.ibm.com>
Acked-by: Dinar Valeev <dvaleev@suse.com>

--- b/arch/powerpc/platforms/pseries/dlpar.c
+++ a/arch/powerpc/platforms/pseries/dlpar.c
@@ -421,6 +421,12 @@ static ssize_t dlpar_cpu_probe(const cha
 		goto out;
 	}
 
+	rc = dlpar_acquire_drc(drc_index);
+	if (rc) {
+		rc = -EINVAL;
+		goto out;
+	}
+
 	dn = dlpar_configure_connector(drc_index);
 	if (!dn) {
 		rc = -EINVAL;
@@ -441,13 +447,6 @@ static ssize_t dlpar_cpu_probe(const cha
 	kfree(dn->full_name);
 	dn->full_name = cpu_name;
 
-	rc = dlpar_acquire_drc(drc_index);
-	if (rc) {
-		dlpar_free_cc_nodes(dn);
-		rc = -EINVAL;
-		goto out;
-	}
-
 	rc = dlpar_attach_node(dn);
 	if (rc) {
 		dlpar_release_drc(drc_index);
